### ISSUE TITLE
SHUTDOWN is in progress

### ISSUE THREAD

**Issue by okmkey45** [Microsoft/msphpsql#423]
Hi

I´ve been having some troubles with a scheduled task in my application. I have a daemon checking the database to trigger possible jobs. A few days ago I experienced a really big peak of errors. Checking out the log, I found three different errors.

First error (thrown two times):

[2017-05-30 12:27:08] production.ERROR: PDOException: SQLSTATE[42000]: [Microsoft][ODBC Driver 13 for SQL Server][SQL Server]SHUTDOWN is in progress. in /var/www/html/flowsoft/vendor/laravel/framework/src/Illuminate/Database/Connection.php:576

Second error (thrown between four and seven times):

[2017-05-30 12:27:08] production.ERROR: PDOException: SQLSTATE[08S01]: [Microsoft][ODBC Driver 13 for SQL Server]TCP Provider: Error code 0x68 in /var/www/html/flowsoft/vendor/laravel/framework/src/Illuminate/Database/Connection.php:576

Third error (since the second error, this error was thrown until 14 thousand times, because of the daemon):

[2017-05-30 12:27:12] production.ERROR: PDOException: SQLSTATE[08S01]: [Microsoft][ODBC Driver 13 for SQL Server]Communication link failure in /var/www/html/flowsoft/vendor/laravel/framework/src/Illuminate/Database/Connection.php:576

What does "SHUTDOWN is in progress" mean? and the other errors? why is this happening? is there a way to control this or maybe change my environment setup?

**Comment by v-kaywon** [comment]
@okmkey45 , is this the first time that it happens? When the error message includes "[SQL Server]", it means the error message comes from SQL Server. It seems like someone has initiated a shutdown progress for your SQL Server (see [here](https://social.msdn.microsoft.com/Forums/sqlserver/en-US/b9a7b2fb-0f41-4eef-a183-70c8a4f6c9c9/shutdown-in-progress-on-ms-sql-2008?forum=sqlsecurity)) and your SQL Server has not been restarted properly. Try to restart your SQL Server and connect again to see if it fixes the issue.
The other two errors are most likely due to you SQL Server shutting down as well. For more information, you can check out the [ODBC Error Codes](https://docs.microsoft.com/en-us/sql/odbc/reference/appendixes/appendix-a-odbc-error-codes).

**Comment by okmkey45** [comment]
No, it is not the first time that this happens. Checking the sql server log I didn´t find anything about the shut down. Do you know how can I check what is initiating the shutdown process?

Also, I have to say. I use a process control system that controls several automated tasks. This system tries to connect with the db continually because I register the tasks in the DB. Apparently, when the shutdown is present, the control system throws this peak of errors.

**Comment by ulvii** [comment]
Hi @okmkey45,

Have you tried checking SQL Server error logs? 

From the article [above](https://social.msdn.microsoft.com/Forums/sqlserver/en-US/b9a7b2fb-0f41-4eef-a183-70c8a4f6c9c9/shutdown-in-progress-on-ms-sql-2008?forum=sqlsecurity): _A shutdown may be initiated by a person or by a program such as cluster administrator, or the shutdown may be the result of problems such as an unhandled exception._

If you are sure that, the shutdown was not initiated by a person, I would suggest to try the following:

- Check cluster logs and errors 
- Check if exceptions are handled properly. According to [this](https://msdn.microsoft.com/en-us/library/8xx3tyca.aspx) article: _When connection pooling is enabled, and if a timeout error or other login error occurs, an exception will be thrown and subsequent connection attempts will fail for the next five seconds, the "blocking period". If the application attempts to connect within the blocking period, the first exception will be thrown again. Subsequent failures after a blocking period ends will result in a new blocking periods that is twice as long as the previous blocking period, up to a maximum of one minute._

You can try to disable connection pooling to see if the issue still persists. 

```
$options = array("ConnectionPooling"=>0); 
$conn = new PDO( "sqlsrv:Server=$server;Database=db", "user", "password", $options); 
```

**Comment by okmkey45** [comment]
Hello. Thanks for the help.

Maybe I should say that I use [supervisord](supervisord.org) for my queues and jobs. Now, this is the weird part. When those errors ocurred, is like the supervisord just tries and tries to connect to the database and those errors are thrown. And I have to restart the supervisor service in order to overcome this. But my application works fine. Should it be something about supervisord?

**Comment by mnaberez** [comment]
> When those errors ocurred, is like the supervisord just tries and tries to connect to the database and those errors are thrown. 

`supervisord` does not contain any code that connects to a database.  Also, the [`PDOException` shown above](https://github.com/Microsoft/msphpsql/issues/423#issue-233588067) is a PHP exception.  `supervisord` is a Python program.  

> Should it be something about supervisord?

It is unlikely.  `supervisord` spawns programs with `fork`/`exec`, and then those programs do whatever they do.  `supervisord` doesn't have a way to cause a `PDOException` or otherwise influence the behavior of the programs after `exec`.

**Comment by yitam** [comment]
Hi @okmkey45 just wondering if you still have issues with this. Thanks.

**Comment by okmkey45** [comment]
Hi @yitam
No, I couldn´t find a solution. 

If you are using Laravel and supervisor, I can tell you that I solved my situation (not a pretty way) by catching all possibles errors regarding this problems (PDOException, shutdown in progress, etc) and restarting automatically the supervisor by a shell command executed from PHP.

Sorry I can´t help you with a better solution
